### PR TITLE
Fix #2986 - Make ModelMetadata getters null-safe

### LINKED ISSUE #2986: `DefaultModelMetadata.PropertyGetter` is not `null`-safe
No current MVC code uses `DefaultModelMetadata.PropertyGetter`. If it were used, would find `null` values e.g. `ModelMetadata.PropertyGetter(null)` lead to NREs. Root cause may be the fact `PropertyHelper.MakeNullSafeFastPropertyGetter()` is not referenced.

Experienced this issue while developing PR #2984.

### PR DESCRIPTION
**Pull request by rynowak**
/cc @dougbu

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs [hunk 1]
```diff
@@ -122,7 +122,8 @@ protected virtual DefaultMetadataDetails[] CreatePropertyDetails([NotNull] Model
                 var propertyEntry = new DefaultMetadataDetails(propertyKey, attributes);
                 if (propertyHelper.Property.CanRead && propertyHelper.Property.GetMethod?.IsPublic == true)
                 {
-                    propertyEntry.PropertyGetter = propertyHelper.ValueGetter;
+                    var getter = PropertyHelper.MakeNullSafeFastPropertyGetter(propertyHelper.Property);
+                    propertyEntry.PropertyGetter = getter;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: Why isn't `propertyHelper.ValueGetter` `null`-safe? Do we need the `Make*()` methods or the `null`-unsafe bits at all?

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs [hunk 2]
```diff
@@ -287,10 +287,7 @@ private static bool CanUpdateReadOnlyProperty(Type propertyType)
                 // instances or overwrite inner properties that are not bound. No need for this with simple values
                 // because they will be overwritten if binding succeeds. Arrays are never reused because they cannot
                 // be resized.
-                //
-                // ModelMetadata.PropertyGetter is not null safe; use it only if Model is non-null.
-                if (bindingContext.Model != null &&
-                    propertyMetadata.PropertyGetter != null &&
+                if (propertyMetadata.PropertyGetter != null &&
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/MutableObjectModelBinder.cs]: :dancers:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: Should test `null` safety in `DefaultModelMetadataProviderTest`. At the moment nothing except product code (`MutableObjectModelBinder` i.e. no test code) does anything beyond check that `metadata.PropertyGetter` is itself non-`null`.

**Comment by dougbu** [comment]
Design looks good. Ready for the real PR :smiling_imp:

File: src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/ModelBinding/Metadata/DefaultModelMetadataProvider.cs]: Will add a test that actually uses the accessors.

As for propertyhelper, I don't see a reason to change it. In the validation/binding code we have a good reason to care about null-safety. In general it's something I'd avoid.

**Comment by rynowak** [comment]
Added a test :+1:

**Comment by dougbu** [comment]
:shipit:

**Pull request merged**